File: sample-java/src/main/java/com/mapp/engagesample/inbox/InboxMessagesAdapter.java
package com.mapp.engagesample.inbox;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncDifferConfig;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.appoxee.shared.InboxMessage;
import com.bumptech.glide.Glide;

import java.util.Objects;

import eu.brrm.shared_ui.databinding.RowInboxMessageBinding;

public class InboxMessagesAdapter extends ListAdapter<InboxMessage, InboxMessagesAdapter.MessagesViewHolder> {

    public interface InboxActionsCallback {
        void onClick(InboxMessage message, int index);

        void onLongClick(InboxMessage message, int index);
    }

    private static final AsyncDifferConfig<InboxMessage> config = new AsyncDifferConfig.Builder<>(new DiffUtil.ItemCallback<InboxMessage>() {
        @Override
        public boolean areItemsTheSame(@NonNull InboxMessage oldItem, @NonNull InboxMessage newItem) {
            return oldItem.getTemplateId() == newItem.getTemplateId() &&
                    Objects.equals(oldItem.getStatus().getStatus(), newItem.getStatus().getStatus());
        }

        @Override
        public boolean areContentsTheSame(@NonNull InboxMessage oldItem, @NonNull InboxMessage newItem) {
            return Objects.equals(oldItem, newItem);
        }
    }).build();

    private InboxActionsCallback callback;

    protected InboxMessagesAdapter() {
        super(config);
    }

    public void setCallback(InboxActionsCallback callback) {
        this.callback = callback;
    }

    @NonNull
    @Override
    public MessagesViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return MessagesViewHolder.create(parent);
    }

    @Override
    public void onBindViewHolder(@NonNull MessagesViewHolder holder, int position) {
        holder.bind(getItem(position));
        holder.itemView.setOnLongClickListener(v -> {
            callback.onLongClick(getItem(position), position);
            return false;
        });
        holder.itemView.setOnClickListener(v -> {
            callback.onClick(getItem(position), position);
        });
    }

    public static class MessagesViewHolder extends RecyclerView.ViewHolder {

        public MessagesViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public void bind(InboxMessage message) {
            RowInboxMessageBinding binding = RowInboxMessageBinding.bind(itemView);
            binding.tvTitle.setText(message.getSubject());
            binding.tvContent.setText(message.getSummary());
            binding.tvStatus.setText(message.getStatus().getStatus());

            Glide.with(binding.ivImage)
                    .load(message.getIconUrl())
                    .fallback(android.R.drawable.ic_menu_report_image)
                    .error(android.R.drawable.stat_notify_error)
                    .placeholder(android.R.drawable.ic_menu_report_image)
                    .into(binding.ivImage);

        }

        public static MessagesViewHolder create(@NonNull ViewGroup parent) {
            RowInboxMessageBinding binding = RowInboxMessageBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new MessagesViewHolder(binding.getRoot());
        }
    }
}

File: sample-kotlin/src/main/java/com/mapp/engagesample/inbox/InboxMessagesAdapter.kt
package com.mapp.engagesample.inbox

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import coil.load
import coil.transform.RoundedCornersTransformation
import com.appoxee.internal.model.response.inbox.InboxMessage
import eu.brrm.shared_ui.databinding.RowInboxMessageBinding

class InboxMessagesAdapter(
    private val onClick: ((InboxMessage, Int) -> Unit)? = null,
    private val onLongClick: ((InboxMessage, Int) -> Unit)? = null
) :
    ListAdapter<InboxMessage, MessageViewHolder>(InboxMessageDiffCallback()) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        return MessageViewHolder.create(parent)
    }

    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        holder.bind(getItem(position))
        holder.itemView.setOnClickListener {
            onClick?.invoke(getItem(position), position)
        }
        holder.itemView.setOnLongClickListener {
            onLongClick?.invoke(getItem(position), position)
            false
        }
    }
}


class InboxMessageDiffCallback() : DiffUtil.ItemCallback<InboxMessage>() {
    override fun areItemsTheSame(oldItem: InboxMessage, newItem: InboxMessage): Boolean {
        return oldItem.templateId == newItem.templateId && oldItem.status == newItem.status
    }

    override fun areContentsTheSame(oldItem: InboxMessage, newItem: InboxMessage): Boolean {
        return oldItem == newItem
    }
}

class MessageViewHolder(private val binding: RowInboxMessageBinding) : ViewHolder(binding.root) {
    fun bind(message: InboxMessage) {
        binding.tvTitle.text = message.subject
        binding.tvContent.text = message.summary
        binding.tvStatus.text = message.status.status
        binding.ivImage.load(message.iconUrl) {
            crossfade(true)
            transformations(RoundedCornersTransformation(10f))
        }
    }

    companion object {
        fun create(parent: ViewGroup): MessageViewHolder {
            val inflater = LayoutInflater.from(parent.context)
            val binding = RowInboxMessageBinding.inflate(inflater, parent, false)
            return MessageViewHolder(binding)
        }
    }
}